### PR TITLE
Feat/predicate iterate

### PR DESCRIPTION
**Pull request by ccamel**
This PR leverages the lazy sequence support for delayed promises introduced in [axone-protocol/prolog#10](https://github.com/axone-protocol/prolog/pull/10) to refactor the _balance-related_ predicates in the `logic` module.

```prolog
bank_balances(X, Y).
bank_spendable_balances(X, Y).
bank_locked_balances(X, Y).
```

Previously, the implementation fetched all addresses from the blockchain at once, which risked failing due to gas limits. The new approach iterates over addresses as each solution branch is explored, aligning better with Prolog’s natural mechanisms and optimizing resource usage.

## Implementation Notes

- While the addresses are now iterated, the balances for each account are still retrieved in full. Although the token amounts per account are expected to be small, I believe the [Zero One Infinity](https://en.wikipedia.org/wiki/Zero_one_infinity_rule) principle should apply, suggesting that balances should also be iterated. However, this would require a more complex implementation, and it’s worth evaluating whether the effort is justified.
- For simplicity, the iteration leverages the keeper methods originally designed for the gRPC API, effectively making the logic act as a client (consumer). This adds a minor overhead due to the round-trip encoding and decoding of protobufs.

## Test

```sh
> make chain-init
> make chain-start

> axoned query logic ask --program "get_all_balances(List) :- findall(balance(X, Y), bank_balances(X, Y), List)." "get_all_balances(L)."  --node http://localhost:26657
answer:
  has_more: false
  results:
  - error: ""
    substitutions:
    - expression: '[balance(axone1p8u47en82gmzfm259y6z93r9qe63l25d858vqu,[uaxone-999000000]),balance(axone1fl48vsnmsdzcv85q5d2q4z5ajdha8yu3ka9cd2,[uaxone-1000000]),balance(axone1tygms3xhhs3yv487phx3dw4a95jn7t7lzaefm7,[]),balance(axone10d07y265gmmuvt4z0w9aw880jnsr700jeyljzw,[]),balance(axone1jv65s3grqf6v6jl3dp4t6c9t9rk99cd8yegtaz,[uaxone-1922]),balance(axone1m3h30wlvsf8llruxtpukdvsy0km2kum8tht0na,[]),balance(axone17xpfvakm2amg962yls6f84z3kell8c5lnl0hlq,[])]'
      variable: L
  variables:
  - L
gas_used: "8253"
height: "62"
user_output: ""

> axoned query auth accounts --node http://localhost:26657 --output json | jq -r '.accounts.[].value.address' | xargs -I {} sh -c 'echo "{}:" && axoned query bank balances {} --node http://localhost:26657 --output json | jq "del(.pagination)"'
{
  "balances": [
    {
      "denom": "uaxone",
      "amount": "999000000"
    }
  ]
}
axone1fl48vsnmsdzcv85q5d2q4z5ajdha8yu3ka9cd2:
{
  "balances": [
    {
      "denom": "uaxone",
      "amount": "1000000"
    }
  ]
}
axone1tygms3xhhs3yv487phx3dw4a95jn7t7lzaefm7:
{
  "balances": []
}
axone10d07y265gmmuvt4z0w9aw880jnsr700jeyljzw:
{
  "balances": []
}
axone1jv65s3grqf6v6jl3dp4t6c9t9rk99cd8yegtaz:
{
  "balances": [
    {
      "denom": "uaxone",
      "amount": "1922"
    }
  ]
}
axone1m3h30wlvsf8llruxtpukdvsy0km2kum8tht0na:
{
  "balances": []
}
axone17xpfvakm2amg962yls6f84z3kell8c5lnl0hlq:
{
  "balances": []
}
...
```

## Release Notes

- **New Features**
	- Introduced a new `AuthQueryService` for enhanced account query capabilities.
	- Added mock implementations for testing the `AuthQueryService`.
	- Enhanced the context management with new keys for `InterfaceRegistry` and `AuthQueryService`.

- **Improvements**
	- Improved error handling and clarity in the `QueryInterpreter` function.
	- Updated test cases to cover various scenarios for account and bank functionalities.

- **Bug Fixes**
	- Removed outdated methods from mock interfaces to streamline functionality.

- **Documentation**
	- Enhanced comments and documentation for clarity in function usage and parameters.

### CODE REVIEW

File: x/logic/keeper/keeper.go [hunk 1]
```diff
@@ -15,25 +16,22 @@ import (
...
 
-func NewKeeper(
-	cdc codec.BinaryCodec,
-	storeKey,
-	memKey storetypes.StoreKey,
-	authority sdk.AccAddress,
-	authKeeper types.AccountKeeper,
-	bankKeeper types.BankKeeper,
+func NewKeeper(cdc codec.BinaryCodec, interfaceRegistry cdctypes.InterfaceRegistry, storeKey, memKey storetypes.StoreKey,
+	authority sdk.AccAddress, authKeeper types.AccountKeeper, authQueryService types.AuthQueryService, bankKeeper types.BankKeeper,
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/keeper/keeper.go:34]: No, that's correct. The calls to `NewKeeper` are either those belonging to another package (and therefore with different arguments), or are correctly instantiated.

File: x/logic/predicate/util.go [hunk 2]
```diff
@@ -18,23 +23,76 @@ func SortBalances(balances sdk.Coins) {
...
+	return func() (V, bool) {
+		u, ok := next()
+		if !ok {
+			var zeroV V
+			return zeroV, false
+		}
+
+		return f(u), true
+	}
+}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/util.go:37]: Nope.
`IterMap` is meant to transform the output of an iterator using a mapping function. According to the [Single Responsibility Principle (SRP)](https://en.wikipedia.org/wiki/Single-responsibility_principle), adding context handling would introduce an additional responsibility that isn’t aligned with its core purpose. I believe the responsibility for handling context cancellation should fall on the caller or the function providing the next function. If context management is important, the caller can ensure the next function respects the context, keeping `IterMap` focused purely on transformation logic, while context handling is managed separately.

File: x/logic/predicate/util.go [hunk 3]
```diff
@@ -18,23 +23,76 @@
...
+			finished = true
+			return lo.Tuple2[sdk.AccountI, error]{A: nil, B: nil}, false
+		}
+
+		key = res.Pagination.NextKey
+
+		interfaceRegistry := ctx.Value(types.InterfaceRegistryContextKey).(cdctypes.AnyUnpacker)
+		var account sdk.AccountI
+		if err := interfaceRegistry.UnpackAny(res.Accounts[0], &account); err != nil {
+			return lo.Tuple2[sdk.AccountI, error]{A: nil, B: err}, true
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/util.go:72]: I quite agree with the principle, but less so with the suggestion. I believe the better approach is to pass the value explicitly through a function argument, rather than indirectly through the context. This makes the dependency clearer and eliminates the need for type assertions. I will make this change.

File: x/logic/predicate/util.go [hunk 4]
```diff
@@ -18,23 +23,76 @@
...
+		}
+
+		return lo.Tuple2[sdk.AccountI, error]{A: account, B: nil}, true
+	}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/util.go:76]: Yup. That's correct.

File: x/logic/testutil/expected_keepers_mocks.go [hunk 5]
```diff
@@ -9,7 +9,7 @@ import (
 	reflect "reflect"
 
 	types "github.com/cosmos/cosmos-sdk/types"
-	types0 "github.com/cosmos/cosmos-sdk/x/bank/types"
+	types0 "github.com/cosmos/cosmos-sdk/x/auth/types"
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/testutil/expected_keepers_mocks.go:12]: This is generated code. I can control the imports to some extent, but it forces changes to the IC, and I’m not sure it’s necessary.

File: x/logic/predicate/bank_test.go [hunk 6]
```diff
@@ -430,13 +449,44 @@
...
-						WithValue(types.BankKeeperContextKey, bankKeeper)
+						WithValue(types.BankKeeperContextKey, bankKeeper).
+						WithValue(types.AuthKeeperContextKey, accountKeeper).
+						WithValue(types.AuthQueryServiceContextKey, authQueryServiceKeeper).
+						WithValue(types.InterfaceRegistryContextKey, encCfg.InterfaceRegistry)
 					sdk.GetConfig().SetBech32PrefixForAccount("axone", "axonepub")
 
-					Convey("and a bank keeper initialized with the preconfigured balances", func() {
+					Convey("and a bank keeper initialized with the preconfigured balances", func(c C) {
+						addresses := lo.Uniq(
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/bank_test.go:466]: I see the point, but I believe the global variable cannot be circumvented as it’s part of a third-party library. While saving and restoring the state might not fully prevent race conditions or isolate tests, it seems more practical to consider this state as shared across all tests.

File: app/app.go [hunk 7]
```diff
@@ -568,10 +568,12 @@ func New(
...
 	app.LogicKeeper = *logicmodulekeeper.NewKeeper(
 		appCodec,
+		app.interfaceRegistry,
 		keys[logicmoduletypes.StoreKey],
 		keys[logicmoduletypes.MemStoreKey],
 		authtypes.NewModuleAddress(govtypes.ModuleName),
 		app.AccountKeeper,
+		authkeeper.NewQueryServer(app.AccountKeeper),
 		app.BankKeeper,
 		app.provideFS,
```
[the remark this answers is not in the document]
Comment (ccamel) [reply app/app.go:578]: Nope. `NewKeeper()` expects a `types.AuthQueryService`, which is a narrowed interface that only exposes the necessary methods. Naturally, `authzkeeper.NewQueryServer(app.AuthzKeeper)` complies with this interface.

Btw code compiles well.

File: app/app.go [hunk 8]
```diff
@@ -568,10 +568,12 @@
...
```
[the remark this answers is not in the document]
Comment (ccamel) [reply app/app.go:578]: Nope. Function signature perfectly matches.

File: x/logic/predicate/util_test.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+				Convey("Then next() returns an error", func() {
+					result, ok := next()
+					So(ok, ShouldBeTrue)
+					So(result.A, ShouldBeNil)
+					So(result.B, ShouldBeError, errors.New("I was told not to unpack anything"))
+				})
+			})
+		})
+	})
+}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/util_test.go:93]: Honestly, this test was mainly to bump up the coverage a bit, since the function is already well covered in other tests. But yeah, I’ll put in some effort to rework it and make the edge cases clearer.

File: x/logic/predicate/bank_test.go [hunk 10]
```diff
@@ -148,7 +167,7 @@ func TestBank(t *testing.T) {
 				balances:   []bank.Balance{},
 				query:      `bank_balances('foo', X).`,
 				wantResult: []testutil.TermResults{{"X": "[uaxone-100]"}},
-				wantError: fmt.Errorf("error(resource_error(resource_module(bank)),[%s],bank_balances/2)",
+				wantError: fmt.Errorf("error(domain_error(encoding(bech32),foo),[%s],bank_balances/2)",
 					strings.Join(strings.Split("decoding bech32 failed: invalid bech32 string length 3", ""), ",")),
 			},
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/bank_test.go:172]: Alright, I know we’ve discussed this before, and I get the sense you’re not a fan of this approach 😅 (which is understandable). From what I remember, this transformation was initially implemented because error messages used to be atoms, but now they’re strings (essentially a list of characters). The change was made to update the tests without too much hassle.

That said, I agree it would be better to stick with the string as it is, without any transformation. However, I don’t want to address this here since it touches on parts of the code that fall outside the scope of this PR. So, could you create a GitHub issue for us to handle later?

File: x/logic/predicate/util.go [hunk 11]
```diff
@@ -18,23 +23,76 @@
...
+}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/util.go:77]: I understand the suggestion, but I don’t agree in this case. Since we’re dealing "queries" inside the blockchain, each query consumes gas. If we increase the pagination limit to, say, 100 accounts but only need 1, we’ll end up consuming 100 times more gas than necessary. So, the current limit is intentional to optimize for gas usage.

File: x/logic/predicate/util.go [hunk 12]
```diff
@@ -18,23 +23,76 @@
...
+
 // AllBalancesSorted returns the list of balances for the given address, sorted by coin denomination.
-func AllBalancesSorted(sdkContext sdk.Context, bankKeeper types.BankKeeper, bech32Addr sdk.AccAddress) sdk.Coins {
-	fetchedBalances := bankKeeper.GetAllBalances(sdkContext, bech32Addr)
+func AllBalancesSorted(ctx context.Context, bankKeeper types.BankKeeper, bech32Addr sdk.AccAddress) sdk.Coins {
+	fetchedBalances := bankKeeper.GetAllBalances(ctx, bech32Addr)
 	SortBalances(fetchedBalances)
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/util.go:82]: No, the code is correct. Input parameters should be as general as possible unless there’s a specific need for them to be more specialized, which isn’t the case here. That's why I made the change.

File: x/logic/util/prolog.go [hunk 13]
```diff
@@ -52,19 +53,30 @@
 	if callErr != nil {
 		if sdkmath.NewUint(uint64(len(results))).LT(solutionsLimit) {
 			// error is not part of the look-ahead and should be included in the solutions
-			sdkCtx := sdk.UnwrapSDKContext(ctx)
+			if errors.Is(callErr, types.LimitExceeded) {
+				return nil, callErr
+			}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/util/prolog.go:58]: 🤔 I don't think so. We need to differentiate between errors that are part of the look-ahead and those that are not.

File: x/logic/util/prolog.go [hunk 14]
```diff
@@ -52,19 +53,30 @@
...
-				return nil, callErr
-			case errors.As(callErr, &panicErr) && errors.Is(panicErr.OriginErr, engine.ErrMaxVariables):
+			if errors.As(callErr, &panicErr) && errors.Is(panicErr.OriginErr, engine.ErrMaxVariables) {
 				return nil, errorsmod.Wrapf(types.LimitExceeded, panicErr.OriginErr.Error())
-			case sdkCtx.GasMeter().IsOutOfGas():
-				return nil, errorsmod.Wrapf(
-					types.LimitExceeded, "out of gas: %s <%s> (%d/%d)",
-					types.ModuleName, callErr.Error(), sdkCtx.GasMeter().GasConsumed(), sdkCtx.GasMeter().Limit())
 			}
+
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/util/prolog.go:64]: I don't entirely agree. Change doesn't seem to me to convey the same semantics.

File: x/logic/predicate/bank.go [hunk 15]
```diff
@@ -98,79 +80,89 @@
...
-		allBalances := bankKeeper.GetAccountsBalances(sdkContext)
-		promises := make([]func(ctx context.Context) *engine.Promise, 0, len(allBalances))
-		for _, balance := range allBalances {
-			address := balance.Address
-			bech32Addr, err = sdk.AccAddressFromBech32(address)
-			if err != nil {
-				return engine.Error(prolog.WithError(engine.ResourceError(prolog.ResourceModule("bank"), env), err, env))
+				return engine.Unify(vm, CoinsToTerm(coins), balances, cont, env)
+			default:
+				return engine.Error(engine.TypeError(prolog.AtomTypeAtom, address, env))
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/bank.go:164]: The tests don’t cover the scenario where the address isn’t an `atom` because the continuation function is invoked by the `account` predicate, which guarantees that the `address` is always an `atom`. It’s not feasible to write a test for this case unless we _mock_ the `account` predicate to behave incorrectly, but honestly, that feels excessive.

[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/bank.go:164]: Also, keep in mind that we’re using CI with test coverage enabled, so there’s no need for you to check it.

File: x/logic/predicate/bank.go [hunk 16]
```diff
@@ -98,79 +80,89 @@
...
+//
+//	account(?Address)
+//
+// where:
+//   - Address represents the account address (in Bech32 format).
+//
+// # Examples:
+//
+//	# Query the locked coins of the account.
+//	- account('axone1ffd5wx65l407yvm478cxzlgygw07h79sw4jwpa', X).
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/bank.go:99]: Oh, yes! Good spot! 👍

**Comment by ccamel** [comment]
@coderabbitai pause

**Comment by ccamel** [comment]
@coderabbitai resolve

I’ve applied the changes we talked about.

**Review by amimart — APPROVED** [review]
Nice one thanks!
